### PR TITLE
+ across(names=)

### LINKED ISSUE #4781: across() should use glue syntax to name output
This resolves a few problems:

* it makes it easier to use multiple sets of `across()` on the same variables without name clashes
* you can put the function names before the variable names if so desired
* the average user is less likely to be unnecessarily exposed to df-cols

(this wouldn't give you the ability to change the order of the variables, which judging from tidyr issues, people will want)

### PR DESCRIPTION
**Pull request by romainfrancois**
closes #4781 

``` r
library(dplyr, warn.conflicts = FALSE)

iris %>%
  group_by(Species) %>%
  summarise(across(starts_with("Sepal"), mean, names = "mean_{col}"))
#> # A tibble: 3 x 3
#>   Species    mean_Sepal.Length mean_Sepal.Width
#>   <fct>                  <dbl>            <dbl>
#> 1 setosa                  5.01             3.43
#> 2 versicolor              5.94             2.77
#> 3 virginica               6.59             2.97
iris %>%
  group_by(Species) %>%
  summarise(across(starts_with("Sepal"), mean, names = "{fn}_{col}"))
#> Error: {fn} cannot be used in the single function case

iris %>%
  group_by(Species) %>%
  summarise(across(starts_with("Sepal"), list(mean = mean, sd = sd), names = "{col}.{fn}"))
#> # A tibble: 3 x 5
#>   Species  Sepal.Length.me… Sepal.Length.sd Sepal.Width.mean Sepal.Width.sd
#>   <fct>               <dbl>           <dbl>            <dbl>          <dbl>
#> 1 setosa               5.01           0.352             3.43          0.379
#> 2 versico…             5.94           0.516             2.77          0.314
#> 3 virgini…             6.59           0.636             2.97          0.322

iris %>%
  group_by(Species) %>%
  summarise(across(starts_with("Sepal"), list(mean = mean, sd), names = "{col}.{fn}"))
#> # A tibble: 3 x 5
#>   Species    Sepal.Length.me… Sepal.Length.2 Sepal.Width.mean Sepal.Width.2
#>   <fct>                 <dbl>          <dbl>            <dbl>         <dbl>
#> 1 setosa                 5.01          0.352             3.43         0.379
#> 2 versicolor             5.94          0.516             2.77         0.314
#> 3 virginica              6.59          0.636             2.97         0.322
    
iris %>%
  group_by(Species) %>%
  summarise(across(starts_with("Sepal"), list(mean, sd), names = "{col}.fn{fn}"))
#> # A tibble: 3 x 5
#>   Species Sepal.Length.fn1 Sepal.Length.fn2 Sepal.Width.fn1 Sepal.Width.fn2
#>   <fct>              <dbl>            <dbl>           <dbl>           <dbl>
#> 1 setosa              5.01            0.352            3.43           0.379
#> 2 versic…             5.94            0.516            2.77           0.314
#> 3 virgin…             6.59            0.636            2.97           0.322
```

<sup>Created on 2020-01-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

### CODE REVIEW

File: tests/testthat/test-across.R [hunk 1]
```diff
@@ -4,3 +4,108 @@ test_that("across() does not select grouping variables", {
   out <- df %>% group_by(g) %>% summarise(x = across(everything())) %>% pull()
   expect_equal(out, tibble(x = 1))
 })
+
+test_that("across() names the output after the selected columns by default in the single fn case", {
+  expect_identical(
+    iris %>%
```
Comment (hadley) [inline tests/testthat/test-across.R]: Can you please use a dummy dataset and `expect_named()`?

File: tests/testthat/test-across.R [hunk 2]
```diff
@@ -4,3 +4,108 @@ test_that("across() does not select grouping variables", {
...
+      group_by(Species) %>%
+      summarise(
+        mean_Sepal.Length = mean(Sepal.Length),
+        mean_Sepal.Width = mean(Sepal.Width)
+      )
+  )
+})
+
+test_that("across(names=) rejects use of {fn} in the single function case", {
```
Comment (hadley) [inline tests/testthat/test-across.R]: Related to my comments above, I don't think this needs to be an error

File: R/across.R [hunk 3]
```diff
@@ -58,24 +68,49 @@ across <- function(cols = everything(), fns = NULL) {
...
 
   if (is.null(fns)) {
-    data
-  } else if (is.function(fns) || is_formula(fns)) {
-    fns <- as_function(fns)
+    return(data)
+  }
+
+  if (is.function(fns) || is_formula(fns)) {
+    if (is.null(names)) {
```
Comment (hadley) [inline R/across.R]: `names <- names %||% "{col}"` ?

File: R/across.R [hunk 4]
```diff
@@ -58,24 +68,49 @@ across <- function(cols = everything(), fns = NULL) {
...
+  }
+  fns <- map(fns, as_function)
+
+  # set default for names if still NULL
+  if (is.null(names)) {
+    names <- "{col}_{fn}"
+  }
+
+  # main loop
+  cols <- list()
```
Comment (hadley) [inline R/across.R]: Maybe `cols <- vector(length(fns) * ncol(data), "list")` ?

File: R/across.R [hunk 5]
```diff
@@ -58,24 +68,49 @@ across <- function(cols = everything(), fns = NULL) {
...
+  names_cols <- names(data)
+  for (i in seq_along(data)) {
+    data_i <- data[[i]]
+    name_i <- names_cols[i]
+    res <- map(fns, function(f) {
```
Comment (hadley) [inline R/across.R]: I think it'd be a little clearer to make this a double loop? 

Either that or do `expand.grid(names(data), names(funs))` + `map2()` and then a vectorised glue call?

File: tests/testthat/test-across.R [hunk 6]
```diff
@@ -4,3 +4,88 @@ test_that("across() does not select grouping variables", {
   out <- df %>% group_by(g) %>% summarise(x = across(everything())) %>% pull()
   expect_equal(out, tibble(x = 1))
 })
+
+test_that("across() correctly names output columns automatically", {
+  expect_named(
```
Comment (hadley) [inline tests/testthat/test-across.R]: I'd pull out `gf <-  group(tibble(x = 1, y = 2, g = ""), x)` — then I think each test will only need one line for the computation.

File: tests/testthat/test-across.R [hunk 7]
```diff
@@ -4,3 +4,88 @@ test_that("across() does not select grouping variables", {
...
+  )
+  expect_named(
+    tibble(x = 1, y = 2, z = 3, g = "") %>%
+      group_by(x) %>%
+      summarise(across(is.numeric, list(mean, sum))),
+    c("x", "y_1", "y_2", "z_1", "z_2")
+  )
+})
+
+test_that("across() respect names= glue syntax", {
```
Comment (hadley) [inline tests/testthat/test-across.R]: Maybe only one test is necessary here? I don't think you're covering any more paths through the code.

**Comment by romainfrancois** [comment]
Thanks, I simplified both implementation and tests

File: R/across.R [hunk 8]
```diff
@@ -58,24 +68,51 @@ across <- function(cols = everything(), fns = NULL) {
   data <- mask$pick(names(vars))
 
   if (is.null(fns)) {
-    data
-  } else if (is.function(fns) || is_formula(fns)) {
-    fns <- as_function(fns)
+    if (is.null(names)) {
+      return(data)
+    } else {
+      return(set_names(data, glue(names, col = names(data), fn = "1")))
```
Comment (hadley) [inline R/across.R]: 👍

**Comment by romainfrancois** [comment]
Merging this now as the errors are related to https://github.com/r-lib/vctrs/issues/776 and/or https://github.com/tidyverse/glue/issues/166 and have nothing to this.

**Pull request merged**